pkg/controller/windowsmachineconfig/windows/connectivity.go

```diff
@@ -0,0 +1,135 @@
+package windows
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+// sshPort is the default SSH port
+const sshPort = "22"
+
+// connectivity encapsulates the information needed to connect to the Windows VM over ssh
+type connectivity struct {
+	// username is the user to connect to the VM
+	username string
+	// ipAddress is the VM's IP address
+	ipAddress string
+	// signer is used for authenticating against the VM
+	signer ssh.Signer
+	// sshClient is the client used to access the Windows VM via ssh
+	sshClient *ssh.Client
+}
+
+// newConnectivity returns an instance of connectivity
+func newConnectivity(username, ipAddress string, signer ssh.Signer) (*connectivity, error) {
+	c := &connectivity{
+		username:  username,
+		ipAddress: ipAddress,
+		signer:    signer,
+	}
+	err := c.initSSHClient()
+	if err != nil {
+		return nil, errors.Wrap(err, "error instantiating SSH client")
+	}
+	return c, nil
+}
+
+// initSSHClient returns a key based SSH client
+func (c *connectivity) initSSHClient() error {
+	if c.username == "" || c.ipAddress == "" || c.signer == nil {
+		return fmt.Errorf("incomplete connectivity information: %v", c)
+	}
+
+	config := &ssh.ClientConfig{
+		User: c.username,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(c.signer),
```

Comment:
signer is being passed down through many functions and a part of the `windows` struct, to only be used in this way. It might make more sense to create the `ssh.AuthMethod` object further up the chain and pass that down instead, as that is what we are actually interested in.

Reply:
`signer` is also used to get the public key. Please see #76. This is the best intermediate data structure we found that satisfies both requirements.